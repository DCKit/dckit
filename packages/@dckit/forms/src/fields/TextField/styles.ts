import { makeStyles } from '@material-ui/core/styles'

export const useStyles = makeStyles(() => ({
  helperText: {
    position: 'absolute',
    top: 50,
    lineHeight: '1.2em',
    cursor: 'default',
  },
}))
