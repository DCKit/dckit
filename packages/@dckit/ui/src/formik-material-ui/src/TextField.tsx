import * as React from 'react';
import MuiTextField, {
  TextFieldProps as MuiTextFieldProps,
} from '@material-ui/core/TextField';
import {
  FieldInputProps,
  FieldMetaProps,
  FieldHelperProps,
  useField,
  useFormikContext,
} from 'formik';

export type TextFieldProps = Omit<
  MuiTextFieldProps,
  'name' | 'value' | 'error'
> & {
  name: string;
};

export function useFieldToTextField<Val = unknown>(
  { disabled, name, ...props }: TextFieldProps,
  customize?: (
    props: [FieldInputProps<Val>, FieldMetaProps<Val>, FieldHelperProps<Val>]
  ) => Partial<Omit<TextFieldProps, 'name'>>
): MuiTextFieldProps {
  const { isSubmitting } = useFormikContext();
  const fieldProps = useField(name);

  const [field, meta] = fieldProps;

  const fieldError = meta.error;
  const showError = meta.touched && !!fieldError;

  return {
    ...props,
    ...field,
    error: showError,
    helperText: showError ? fieldError : props.helperText,
    disabled: disabled ?? isSubmitting,
    variant: props.variant,
    ...customize?.(fieldProps),
  };
}

export function TextField({ children, ...props }: TextFieldProps) {
  return (
    <MuiTextField {...useFieldToTextField(props)}>{children}</MuiTextField>
  );
}

TextField.displayName = 'FormikMaterialUITextField';
