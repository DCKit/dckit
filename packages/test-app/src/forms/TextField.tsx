import React from 'react'
import {
  TextField as MuiTextField,
  InputAdornment,
  StandardTextFieldProps,
} from '@material-ui/core'
import { IFormField } from './types'
/*
  FormHelperTextProps={{
    classes: {
      root: classes.helperText,
      error: classes.helperTextError,
    },
  }}
  InputProps={{
    endAdornment: (
      <InputAdornment position="start" className={classes.inputSuffix}>
        {inputSuffix}
      </InputAdornment>
    ),
  }}
*/

export const TextField = (props: IFormField) => {
  const handleBlur = (e: any) => {
    const { onBlur, onChange } = props
    const value = String(e.target.value || '')
    const trimValue = value.trim()

    if (value !== trimValue) {
      e.target.value = trimValue
      onChange && onChange(e)
    }
    onBlur && onBlur(e)
  }

  const { initialValue, suffix, size, field, ...restProps } = props
  const fieldProps: StandardTextFieldProps = {
    ...restProps,
    name: field,
    onBlur: handleBlur,
  }
  if (suffix) {
    fieldProps.InputProps = {
      endAdornment: <InputAdornment position="end">{suffix}</InputAdornment>,
    }
  }
  return <MuiTextField {...fieldProps} />
}
