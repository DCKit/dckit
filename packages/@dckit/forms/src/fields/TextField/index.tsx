import React from 'react'
import {
  TextField as MuiTextField,
  InputAdornment,
  InputProps,
  TextFieldProps,
} from '@material-ui/core'
import { MuiFieldProps } from '../../types'
import { useStyles } from './styles'

const Adornment = React.memo(
  ({ position, adornment }: { position: 'start' | 'end'; adornment: any }) => (
    <InputAdornment position={position}>{adornment}</InputAdornment>
  )
)

export const TextField = (props: MuiFieldProps) => {
  const classes = useStyles()

  const { startAdornment, endAdornment, variant, ...restProps } = props

  const fieldProps: TextFieldProps = {
    ...restProps,
    variant,
    fullWidth: true,
    FormHelperTextProps: {
      classes: {
        root:
          variant === 'outlined'
            ? classes.helperTextOutlined
            : classes.helperText,
      },
    },
  }

  if (variant === 'outlined' || variant === 'filled') {
    fieldProps.size = 'small'
  }

  const inputProps: Partial<InputProps> = { ...fieldProps.InputProps }

  if (startAdornment) {
    inputProps.startAdornment = (
      <Adornment position="start" adornment={startAdornment} />
    )
  }
  if (endAdornment) {
    inputProps.endAdornment = (
      <Adornment position="end" adornment={endAdornment} />
    )
  }
  fieldProps.InputProps = inputProps

  return <MuiTextField {...fieldProps} />
}
