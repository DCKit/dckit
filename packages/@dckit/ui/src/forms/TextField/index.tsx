import React from 'react'
import {
  TextField as MuiTextField,
  InputAdornment,
  StandardTextFieldProps,
} from '@material-ui/core'
import { MuiFieldProps } from '../types'
import { useStyles } from './styles'

const Adornment = React.memo(
  ({ position, adornment }: { position: 'start' | 'end'; adornment: any }) => (
    <InputAdornment position={position}>{adornment}</InputAdornment>
  )
)

export const TextField = (props: MuiFieldProps) => {
  const classes = useStyles()

  const { startAdornment, endAdornment, ...restProps } = props

  const fieldProps: StandardTextFieldProps = {
    ...restProps,
    FormHelperTextProps: {
      classes: {
        root: classes.helperText,
      },
    },
  }

  const inputProps = fieldProps.InputProps || {}

  if (startAdornment) {
    inputProps.startAdornment = (
      <Adornment position="start" adornment={startAdornment} />
    )
  }
  if (endAdornment) {
    inputProps.endAdornment = (
      <Adornment position="end" adornment={endAdornment} />
    )
  }
  fieldProps.InputProps = inputProps

  return <MuiTextField {...fieldProps} fullWidth={true} />
}
