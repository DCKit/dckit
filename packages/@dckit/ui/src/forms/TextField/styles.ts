import { Theme } from '@material-ui/core'
import { makeStyles } from '@material-ui/core/styles'

export const useStyles = makeStyles((theme: Theme) => {
  return {
    helperText: {
      position: 'absolute',
      bottom: -18,
    },
  }
})
