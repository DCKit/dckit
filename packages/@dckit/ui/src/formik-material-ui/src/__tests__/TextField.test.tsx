import * as React from 'react';
import { Formik, Form } from 'formik';
import renderer from 'react-test-renderer';

import { TextField } from '../TextField';

test('TextField Renders Correctly', () => {
  const component = renderer.create(
    <Formik initialValues={{}} onSubmit={() => {}}>
      <Form>
        <TextField name="test" label="Text" />
      </Form>
    </Formik>
  );

  expect(component.toJSON()).toMatchSnapshot();
});

test('Outlined TextField', () => {
  const component = renderer.create(
    <Formik initialValues={{}} onSubmit={() => {}}>
      <Form>
        <TextField name="test" label="Text" variant="outlined" />
      </Form>
    </Formik>
  );

  expect(component.toJSON()).toMatchSnapshot();
});
