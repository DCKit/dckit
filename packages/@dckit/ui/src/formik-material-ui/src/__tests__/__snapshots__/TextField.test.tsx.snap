// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Outlined TextField 1`] = `
<form
  action="#"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="MuiFormControl-root MuiTextField-root"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
      data-shrink={false}
    >
      Text
    </label>
    <div
      className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        autoFocus={false}
        className="MuiInputBase-input MuiOutlinedInput-input"
        disabled={false}
        name="test"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        required={false}
        type="text"
      />
      <fieldset
        aria-hidden={true}
        className="PrivateNotchedOutline-root-75 MuiOutlinedInput-notchedOutline"
        style={
          Object {
            "paddingLeft": 8,
          }
        }
      >
        <legend
          className="PrivateNotchedOutline-legend-76"
          style={
            Object {
              "width": 0.01,
            }
          }
        >
          <span
            dangerouslySetInnerHTML={
              Object {
                "__html": "&#8203;",
              }
            }
          />
        </legend>
      </fieldset>
    </div>
  </div>
</form>
`;

exports[`TextField Renders Correctly 1`] = `
<form
  action="#"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="MuiFormControl-root MuiTextField-root"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
      data-shrink={false}
    >
      Text
    </label>
    <div
      className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        autoFocus={false}
        className="MuiInputBase-input MuiInput-input"
        disabled={false}
        name="test"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        required={false}
        type="text"
      />
    </div>
  </div>
</form>
`;
